pkg/issuer/acme/dns/dns.go

```diff
@@ -165,21 +166,28 @@ func (s *Solver) solverForIssuerProvider(providerName string) (solver, error) {
 			return nil, errors.Wrap(err, "error instantiating akamai challenge solver")
 		}
 	case providerConfig.CloudDNS != nil:
-		saSecret, err := s.secretLister.Secrets(s.resourceNamespace).Get(providerConfig.CloudDNS.ServiceAccount.Name)
-		if err != nil {
-			return nil, fmt.Errorf("error getting clouddns service account: %s", err)
-		}
+		if !providerConfig.CloudDNS.UseMetadataAuth {
+			saSecret, err := s.secretLister.Secrets(s.resourceNamespace).Get(providerConfig.CloudDNS.ServiceAccount.Name)
+			if err != nil {
+				return nil, fmt.Errorf("error getting clouddns service account: %s", err)
+			}
 
-		saKey := providerConfig.CloudDNS.ServiceAccount.Key
-		saBytes := saSecret.Data[saKey]
+			saKey := providerConfig.CloudDNS.ServiceAccount.Key
+			saBytes := saSecret.Data[saKey]
 
-		if len(saBytes) == 0 {
-			return nil, fmt.Errorf("specfied key %q not found in secret %s/%s", saKey, saSecret.Namespace, saSecret.Name)
-		}
+			if len(saBytes) == 0 {
+				return nil, fmt.Errorf("specfied key %q not found in secret %s/%s", saKey, saSecret.Namespace, saSecret.Name)
+			}
 
-		impl, err = s.dnsProviderConstructors.cloudDNS(providerConfig.CloudDNS.Project, saBytes)
-		if err != nil {
-			return nil, fmt.Errorf("error instantiating google clouddns challenge solver: %s", err)
+			impl, err = s.dnsProviderConstructors.cloudDNS(providerConfig.CloudDNS.Project, saBytes)
+			if err != nil {
+				return nil, fmt.Errorf("error instantiating google clouddns challenge solver: %s", err)
+			}
+		} else {
+			impl, err = s.dnsProviderConstructors.cloudDNSMetaAuth(providerConfig.CloudDNS.Project)
+			if err != nil {
+				return nil, fmt.Errorf("error instantiating google clouddns meta auth challenge solver: %s", err)
+			}
```

Comment:
Can you adjust this to not touch the previous lines?

Something like:

```go
if providerConfig.CloudDNS.UseMetadataAuth {
	impl, err = s.dnsProviderConstructors.cloudDNSMetaAuth(providerConfig.CloudDNS.Project)
	if err != nil {
		return nil, fmt.Errorf("error instantiating google clouddns meta auth challenge solver: %s", err)
	}
	break
}
...
```

Reply:
Also, instead of using `UseMetadataAuth` to determine whether to use the metadata server, we already have the concept of 'ambient credentials' that can be enabled/disabled for ClusterIssuer or Issuers with a flag on cert-manager itself.

Can you adjust your code to be similar to the Route53 provider in using these? Then, if a user specifies a ServiceAccount we will use that. If they do not specify a ServiceAccount, we will attempt to use the metadata server.